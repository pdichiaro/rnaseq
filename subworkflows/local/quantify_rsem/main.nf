//
// Gene/transcript quantification with RSEM
//

include { RSEM_CALCULATEEXPRESSION } from '../../../modules/nf-core/rsem/calculateexpression'
include { RSEM_MERGE_COUNTS        } from '../../../modules/local/rsem_merge_counts'
include { SENTIEON_RSEMCALCULATEEXPRESSION } from '../../../modules/nf-core/sentieon/rsemcalculateexpression'

workflow QUANTIFY_RSEM {
    take:
    reads             // channel: [ val(meta), [ reads ] ] - FASTQ or BAM files
    index             // channel: /path/to/rsem/index/
    use_sentieon_star // boolean: determines whether RSEM is run with Sentieon accelerated STAR
    annotation_matrix // channel: path(annotation_matrix.txt)

    main:

    ch_versions = Channel.empty()

    //
    // Quantify reads with RSEM
    //
    ch_rsem_out = null
    if (use_sentieon_star){
        SENTIEON_RSEMCALCULATEEXPRESSION ( reads, index )
        ch_rsem_out = SENTIEON_RSEMCALCULATEEXPRESSION
    } else {
        RSEM_CALCULATEEXPRESSION ( reads, index )
        ch_rsem_out = RSEM_CALCULATEEXPRESSION
    }

    ch_counts_gene = ch_rsem_out.out.counts_gene
    ch_counts_transcript = ch_rsem_out.out.counts_transcript
    ch_stat = ch_rsem_out.out.stat
    ch_logs = ch_rsem_out.out.logs
    ch_versions = ch_versions.mix(ch_rsem_out.out.versions.first())

    //
    // Merge counts across samples
    //
    RSEM_MERGE_COUNTS (
        ch_counts_gene.collect{it[1]},       // [meta, counts]: Collect the second element (counts files) in the channel across all samples
        ch_counts_transcript.collect{it[1]},
        annotation_matrix
    )
    ch_versions = ch_versions.mix(RSEM_MERGE_COUNTS.out.versions)

    emit:
    counts_gene              = ch_counts_gene                                 // channel: [ val(meta), counts ]
    counts_transcript        = ch_counts_transcript                           // channel: [ val(meta), counts ]
    stat                     = ch_stat                                        // channel: [ val(meta), stat ]
    logs                     = ch_logs                                        // channel: [ val(meta), logs ]

    merged_counts_gene       = RSEM_MERGE_COUNTS.out.counts_gene              //    path: *.gene_counts.tsv
    merged_tpm_gene          = RSEM_MERGE_COUNTS.out.tpm_gene                 //    path: *.gene_tpm.tsv
    merged_counts_transcript = RSEM_MERGE_COUNTS.out.counts_transcript        //    path: *.transcript_counts.tsv
    merged_tpm_transcript    = RSEM_MERGE_COUNTS.out.tpm_transcript           //    path: *.transcript_tpm.tsv

    versions                 = ch_versions                                    // channel: [ versions.yml ]
}
