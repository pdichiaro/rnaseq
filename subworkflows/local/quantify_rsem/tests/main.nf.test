nextflow_workflow {

    name "Test Workflow QUANTIFY_RSEM"
    script "../main.nf"
    workflow "QUANTIFY_RSEM"
    config "./nextflow.config"

    test("homo_sapiens") {
        tag "test"
        setup {
            run("RSEM_PREPAREREFERENCE") {
                script "../../../../modules/nf-core/rsem/preparereference/main.nf"
                process {
                    """
                    input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    input[1] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true))
                    """
                }
            }
        }

        when {
            workflow {
                """
                use_sentieon_star        = false

                input[0] = Channel.of([
                    [ id:'test', strandedness: 'forward' ], // meta map
                    file('https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/' + 'genomics/homo_sapiens/illumina/bam/rsem.transcript.bam', checkIfExists: true)
                ])
                input[1] = RSEM_PREPAREREFERENCE.out.index
                input[2] = use_sentieon_star
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.counts_gene,
                    workflow.out.counts_transcript,
                    workflow.out.stat,
                    workflow.out.merged_counts_gene,
                    workflow.out.merged_tpm_gene,
                    workflow.out.merged_counts_transcript,
                    workflow.out.merged_tpm_transcript,
                    workflow.out.versions
                ).match()}
            )
        }
    }

    test("homo_sapiens - stub") {

        options "-stub"

        setup {
            run("RSEM_PREPAREREFERENCE") {
                script "../../../../modules/nf-core/rsem/preparereference/main.nf"
                process {
                    """
                    input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    input[1] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true))
                    """
                }
            }
        }

        when {
            workflow {
                """
                use_sentieon_star        = false

                input[0] = Channel.of([
                    [ id:'test', strandedness: 'forward' ], // meta map
                    file('https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/' + 'genomics/homo_sapiens/illumina/bam/rsem.transcript.bam', checkIfExists: true)
                ])
                input[1] = RSEM_PREPAREREFERENCE.out.index
                input[2] = use_sentieon_star
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("homo_sapiens - sentieon - stub") {

        options "-stub"

        setup {
            run("SENTIEON_RSEMPREPAREREFERENCE") {
                script "../../../../modules/nf-core/sentieon/rsempreparereference/main.nf"
                process {
                    """
                    input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    input[1] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true))
                    """
                }
            }
        }

        when {
            workflow {
                """
                use_sentieon_star        = true

                input[0] = Channel.of([
                    [ id:'test', strandedness: 'forward' ], // meta map
                    file('https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/' + 'genomics/homo_sapiens/illumina/bam/rsem.transcript.bam', checkIfExists: true)
                ])
                input[1] = SENTIEON_RSEMPREPAREREFERENCE.out.index
                input[2] = use_sentieon_star
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
