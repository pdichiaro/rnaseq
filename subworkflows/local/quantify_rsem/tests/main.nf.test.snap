{
    "homo_sapiens - sentieon - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.genes.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.isoforms.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.stat:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "rsem.merged.gene_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "5": [
                    "rsem.merged.gene_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "6": [
                    "rsem.merged.transcript_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "7": [
                    "rsem.merged.transcript_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "8": [
                    "versions.yml:md5,2aa5252eb2ffb409cf556a165d40f8a9",
                    "versions.yml:md5,773c15c4ecb7d486a4bdd8ef73e7ac5d"
                ],
                "counts_gene": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.genes.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "counts_transcript": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.isoforms.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "logs": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "merged_counts_gene": [
                    "rsem.merged.gene_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_counts_transcript": [
                    "rsem.merged.transcript_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_tpm_gene": [
                    "rsem.merged.gene_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_tpm_transcript": [
                    "rsem.merged.transcript_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "stat": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.stat:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2aa5252eb2ffb409cf556a165d40f8a9",
                    "versions.yml:md5,773c15c4ecb7d486a4bdd8ef73e7ac5d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-15T16:59:19.581004"
    },
    "homo_sapiens - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.genes.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.isoforms.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.stat:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "rsem.merged.gene_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "5": [
                    "rsem.merged.gene_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "6": [
                    "rsem.merged.transcript_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "7": [
                    "rsem.merged.transcript_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "8": [
                    "versions.yml:md5,2aa5252eb2ffb409cf556a165d40f8a9",
                    "versions.yml:md5,ae1676b60b6335fff8f1188288103a1c"
                ],
                "counts_gene": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.genes.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "counts_transcript": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.isoforms.results:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "logs": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "merged_counts_gene": [
                    "rsem.merged.gene_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_counts_transcript": [
                    "rsem.merged.transcript_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_tpm_gene": [
                    "rsem.merged.gene_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "merged_tpm_transcript": [
                    "rsem.merged.transcript_tpm.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "stat": [
                    [
                        {
                            "id": "test",
                            "strandedness": "forward"
                        },
                        "test.stat:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2aa5252eb2ffb409cf556a165d40f8a9",
                    "versions.yml:md5,ae1676b60b6335fff8f1188288103a1c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-15T16:56:01.229068"
    },
    "homo_sapiens": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.genes.results:md5,c7ec226f76736ea805771e73553ae359"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.isoforms.results:md5,99f7f80aa505b44ca429fdebbd7dd5d8"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    [
                        "test.cnt:md5,76249e6b2f3c104f414aae596ba2c2f4",
                        "test.model:md5,a7a4bc1734918ef5848604e3362b83e2",
                        "test.theta:md5,de2e4490c98cc5383a86ae8225fd0a28"
                    ]
                ]
            ],
            [
                "rsem.merged.gene_counts.tsv:md5,f3b586000c97c89aec45360772e73df1"
            ],
            [
                "rsem.merged.gene_tpm.tsv:md5,e6ae3dfa9913a7c15b705ec27c376212"
            ],
            [
                "rsem.merged.transcript_counts.tsv:md5,235c002b811360183494a2aac74d1f5b"
            ],
            [
                "rsem.merged.transcript_tpm.tsv:md5,d38f22667e3b8012f1d783b8d5159ccc"
            ],
            [
                "versions.yml:md5,2aa5252eb2ffb409cf556a165d40f8a9",
                "versions.yml:md5,ae1676b60b6335fff8f1188288103a1c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-09-16T08:19:00.078928064"
    }
}